torchmetrics/functional/text/helper.py

```diff
@@ -12,7 +12,348 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import List
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+# Copyright 2020 Memsource
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+from enum import Enum, IntEnum, unique
+from typing import Dict, List, Sequence, Tuple, Union
+
+# Tercom-inspired limits
+_BEAM_WIDTH = 25
+
+# Sacrebleu-inspired limits
+_MAX_CACHE_SIZE = 10000
+_INT_INFINITY = int(1e16)
+
+
+@unique
+class _EDIT_OPERATIONS(str, Enum):
+    """Enumerations for the Levenhstein edit operations."""
+
+    OP_INSERT = "insert"
+    OP_DELETE = "delete"
+    OP_SUBSTITUTE = "substitute"
+    OP_NOTHING = "nothing"
+    OP_UNDEFINED = "undefined"
+
+
+class _EDIT_OPERATIONS_COST(IntEnum):
+    """Enumerations for the Levenhstein edit operation costs."""
+
+    OP_INSERT = 1
+    OP_DELETE = 1
+    OP_SUBSTITUTE = 1
+    OP_NOTHING = 0
+
+
+class _LevenshteinEditDistance:
+    """A convenience class for calculating the Levenshtein edit distance, which also caches some intermediate
+    values to hasten the calculation.
+
+    The implementation follows the implemenation from
+    https://github.com/mjpost/sacrebleu/blob/master/sacrebleu/metrics/lib_ter.py, where the most of this implementation
+    is adapted and copied from.
```

Comment:
it is not clear if most are copied or adapted...
this is tiny ice we were once in past and we shall be careful about copy others code :rabbit: 
I may say that the best way would be getting original author approval on using an adaptation of his code...
but I see that his license is also Apache2 so it could be fine :]

Reply:
@Borda Yes, I definitely get your point here. I will go through the review (thanks a lot for that O:]) and try to specify in more details where and how the code is adapted/copied.

Reply:
as long as they use an Apache2 license and we clearly state in the docstring where it is comes from and if it is adapted or not, then we should be fine